proposals/0002-service-discovery.md

```diff
@@ -0,0 +1,626 @@
+
+For the registration part we want to add an integration with the following registries: Consul, Kubernetes, Eureka, Zookeeper.
+For the discovery part we want to have an ability to discover services and instances, and also we want to 
+resolve service names to actual instances, optionally using DNS-style service names 
+(e.g., _service://STORES/product_, where "STORES" is a service name).
+The plugin should be able to integrate with the `HttpClient` to intercept and rewrite such requests 
+based on service registry data.
+
+### Server Plugin
+For a server plugin for Service Discovery we want to have:
+1. Ability to register/deregister services
+2. Ability to discover services or instances
+3. Simple way to configure the registry. And it should have a common way of setup for different registries
+
+### Client Plugin
+For a client plugin for Service Discovery we want to have:
+1. Ability to discover services or instances
+2. Do requests using DNS-style service names (e.g., _service://STORES/product_)
+3. Simple way to configure the registry. And it should have a common way of setup for different registries
+
+To be able to register services or get information about them, we need to have access to the registry. The idea is to
+configure the registry for example like this:
+```kotlin
+install(ServiceDiscovery) { 
+    kubernetes {
+        namespace = "default"
+        masterUrl = "https://kubernetes.default.svc"
+        apiVersion = "v1"
+    }
+    // OR
+    consul { 
+        host = "localhost"
+        port = 8500
+        aclToken = "default"
+        healthCheckUrl = "http://localhost:8080/health"
+    }
+}
+```
+The scheme is illustrated on the image below:
+![img.png](images/0002-service-discovery-scheme.png)
+
+`ServiceRegistry` and `DiscoveryClient` interfaces will declare common methods for all implementations:
+```kotlin
+interface ServiceRegistry {
+    suspend fun register(registration: ServiceInstance)
+    suspend fun deregister(registration: ServiceInstance)
+    // ...
+}
+
+interface DiscoveryClient { 
+    suspend fun getInstances(serviceId: String): List<ServiceInstance>
+    suspend fun getServices(): List<String>
+    // ...
+}
+
+interface ServiceInstance {
+    fun getInstanceId(): String
+    fun getServiceId(): String
+    // ...
+}   
+
+```
+The individual implementations for each Service Discovery solution 
+(e.g., `KubernetesServiceRegistry` and `KubernetesDiscoveryClient`, `ConsulServiceRegistry` and `ConsulDiscoveryClient`, etc.)
+will implement these interfaces and provide the access to functions available to current service registry:
+```kotlin
+class ConsulDiscoveryClient : DiscoveryClient {
+    // common method
+    override suspend fun getInstances(serviceId: String): List<ServiceInstance> { /**...**/ }
+
+    // specific for Consul method
+    fun getConsulInstances(serviceId: String): List<ConsulServiceInstance> { /**...**/ }
```

Comment:
Will it be the same as calling `getInstances(serviceId).map { it as ConsulServiceInstance }`?
Probably we could introduce some generics to `DiscoveryClient` to return instances of the specific type from `getInstances`.
Something like:

```kotlin
interface DiscoveryClient<T : ServiceInstance> {
    suspend fun getInstances(serviceId: String): List<T>
}
```